pkg/network/server.go

```diff
@@ -1654,26 +1668,42 @@ func (s *Server) broadcastTxLoop() {
 	}
 }
 
-// Port returns a server port that should be used in P2P version exchange. In
-// case `AnnouncedPort` is set in the server.Config, the announced node port
-// will be returned (e.g. consider the node running behind NAT). If `AnnouncedPort`
-// isn't set, the port returned may still differs from that of server.Config.
-func (s *Server) Port() (uint16, error) {
-	if s.AnnouncedPort != 0 {
-		return s.ServerConfig.AnnouncedPort, nil
+// Port returns a server port that should be used in P2P version exchange with the
+// peer connected on the given localAddr. In case if announced node port is set
+// in the server.Config for the given bind address, the announced node port will
+// be returned (e.g. consider the node running behind NAT). If `AnnouncedPort`
+// isn't set, the port returned may still differ from that of server.Config. If
+// no localAddr is given, then the first available port will be returned.
+func (s *Server) Port(localAddr net.Addr) (uint16, error) {
+	var connIP string
+	if localAddr != nil {
+		connIP, _, _ = net.SplitHostPort(localAddr.String()) // Ignore error and provide info if possible.
 	}
-	var port uint16
-	_, portStr, err := net.SplitHostPort(s.transport.Address())
-	if err != nil {
-		port = s.ServerConfig.Port
-	} else {
-		p, err := strconv.ParseUint(portStr, 10, 16)
+	for i, tr := range s.transports {
+		listenAddr := tr.Address()
+		// TODO: perform it once during (t *TCPTransport).Accept()?
+		listenIP, listenPort, err := net.SplitHostPort(listenAddr)
 		if err != nil {
-			return 0, err
+			return 0, fmt.Errorf("invalid transporter bind address '%s': %w", listenAddr, err)
+		}
+
+		// TODO: IPv6
+		if listenIP == "::" /* TODO: properly handle case of 0.0.0.0 (listenIP is [::]), i.e. when transporter listens all available addresses */ ||
+			localAddr == nil || connIP == "" || connIP == listenIP {
```

Comment:
Comparing strings is not the best thing here, also you should prioritize direct matches, currently the behavior depends on the order of `s.transports`.

Reply:
Fixed.